README.md
# freedesktop-categories

[![Build Status][s1]][cc] [![Crates.io][s2]][ci] [![Documentation][s3]][docs]

[s1]: https://circleci.com/gh/ebkalderon/freedesktop-categories.svg?style=shield
[cc]: https://circleci.com/gh/ebkalderon/freedesktop-categories
[s2]: https://img.shields.io/crates/v/freedesktop-categories.svg
[ci]: https://crates.io/crates/freedesktop-categories
[s3]: https://img.shields.io/badge/docs-master-blue.svg
[docs]: https://docs.rs/freedesktop-categories

Static hash map of all application categories as defined by the Freedesktop.org
[Desktop Menu Specification][dm].

[dm]: https://specifications.freedesktop.org/menu-spec/menu-spec-latest.html

These categories are used in the parsing of `.desktop` entries on many open
source desktop environments, among other things. They are also present in other
package metadata standards like [AppStream][as].

[as]: https://www.freedesktop.org/software/appstream/docs/index.html

## Contributing

This is a community project that welcomes contributions from anyone. If you're
interested in helping out, please check the [issue tracker][it] to get started.
Pull requests and issues are welcome!

[it]: https://github.com/ebkalderon/freedesktop-categories/issues

## License

`freedesktop-categories` is free and open source software distributed under the
terms of both the [MIT License][lm] and the [Apache License 2.0][la].

[lm]: ./LICENSE-MIT
[la]: ./LICENSE-APACHE

Unless you explicitly state otherwise, any contribution intentionally submitted
for inclusion in the work by you, as defined in the Apache-2.0 license, shall be
dual licensed as above, without any additional terms or conditions.


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [0.2.0] - 2019-02-15
### Added
* Add crate LICENSE file (#5).

### Changed
* Update `map.rs` to reflect the latest spec version.
* Upgrade `phf` dependency in main crate to version 0.7.24.
* Upgrade `amxml` dependency in codegen crate to version 0.5.3.
* Upgrade `curl` dependency in codegen crate to version 0.4.19.
* Upgrade `phf_codegen` dependency in codegen crate to version 0.7.24.

### Fixed
* Update incorrect repo URL in codegen crate (#7).
* Eliminate unused `Result` warning with `writeln!` in codegen crate.

## 0.1.0 - 2018-09-29
### Added
* Initial crate release.
* Static hash map based on the latest Desktop Menu specification.
* Codegen crate for automatic parsing of the upstream DocBook XML.

[Unreleased]: https://github.com/ebkalderon/freedesktop-categories/compare/v0.2.0...HEAD
[0.2.0]: https://github.com/ebkalderon/freedesktop-categories/compare/v0.1.0...v0.2.0


src/lib.rs
//! Static hash map of all application categories defined by the Freedesktop.org
//! [Desktop Menu Specification][dm].
//!
//! [dm]: https://specifications.freedesktop.org/menu-spec/menu-spec-latest.html
//!
//! These categories are used in the parsing of `.desktop` entries on many open source desktop
//! environments, among other things. They are also present in other package metadata standards
//! like [AppStream][as].
//!
//! [as]: https://www.freedesktop.org/software/appstream/docs/index.html

#![deny(missing_debug_implementations)]
#![forbid(unsafe_code)]

extern crate phf;

#[cfg(feature = "generate-map")]
include!(concat!(env!("OUT_DIR"), "/map.rs"));

#[cfg(not(feature = "generate-map"))]
include!("map.rs");

// TODO: Hide away this type and expose something more type-safe.
/// Entry in the category hash map.
#[derive(Clone, Debug)]
pub enum Category {
    Main {
        /// Other main categories required to be used alongside this category.
        requires: &'static [&'static str],
    },
    Additional {
        /// Related categories which are suggested to be used alongside this one.
        suggests: &'static [&'static str],
    },
    /// Reserved tags require users to specify the `OnlyShowIn` tag in entries.
    Reserved,
}

/// Returns whether the given string is a valid category.
///
/// This function does not differentiate between different kinds of categories. Semicolon-delimited
/// category sets such as `Foo;Bar;Baz` are not counted by this function.
///
/// # Example
///
/// ```
/// # use freedesktop_categories::is_category;
/// assert!(is_category("AudioVideo"));
/// assert!(is_category("Office"));
/// assert!(is_category("System"));
/// ```
pub fn is_category(name: &str) -> bool {
    CATEGORIES.contains_key(name)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn main_categories_exist() {
        assert!(is_category("AudioVideo"));
        assert!(is_category("Audio"));
        assert!(is_category("Video"));
        assert!(is_category("Development"));
        assert!(is_category("Education"));
        assert!(is_category("Game"));
        assert!(is_category("Graphics"));
        assert!(is_category("Network"));
        assert!(is_category("Office"));
        assert!(is_category("Settings"));
        assert!(is_category("System"));
        assert!(is_category("Utility"));
    }
}


codegen/src/error.rs
//! Common error type used throughout the crate.

use std::env::VarError;
use std::error::Error as StdError;
use std::fmt::{Display, Formatter, Result as FmtResult};
use std::io::Error as IoError;

use curl::Error as CurlError;

/// Types of errors and their causes.
#[derive(Debug)]
pub enum Error {
    /// Failed to download and/or cache the specification XML file.
    Curl(CurlError),
    /// Failed to read or write from a file descriptor.
    Io(IoError),
    /// Failed to load an environment variable.
    Var(VarError),
    /// Failed to parse the Freedesktop.org Desktop Menu spec.
    Xml(Box<StdError>),
}

impl StdError for Error {
    fn description(&self) -> &str {
        match *self {
            Error::Curl(_) => "Failed to download XML file from Freedesktop.org",
            Error::Io(_) => "Failed to read or write from a file",
            Error::Var(_) => "Failed to load an environment variable",
            Error::Xml(_) => "Could not parse Freedesktop.org Desktop Menu spec",
        }
    }

    fn cause(&self) -> Option<&StdError> {
        match *self {
            Error::Curl(ref e) => Some(e),
            Error::Io(ref e) => Some(e),
            Error::Var(ref e) => Some(e),
            Error::Xml(ref e) => Some(e.as_ref()),
        }
    }
}

impl Display for Error {
    fn fmt(&self, fmt: &mut Formatter) -> FmtResult {
        match *self {
            Error::Curl(ref e) => write!(fmt, "{}: {}", self.description(), e),
            Error::Io(ref e) => write!(fmt, "{}: {}", self.description(), e),
            Error::Var(ref e) => write!(fmt, "{}: {}", self.description(), e),
            Error::Xml(ref e) => write!(fmt, "{}: {}", self.description(), e),
        }
    }
}

impl From<CurlError> for Error {
    fn from(e: CurlError) -> Self {
        Error::Curl(e)
    }
}

impl From<IoError> for Error {
    fn from(e: IoError) -> Self {
        Error::Io(e)
    }
}

impl From<VarError> for Error {
    fn from(e: VarError) -> Self {
        Error::Var(e)
    }
}

impl From<Box<StdError>> for Error {
    fn from(e: Box<StdError>) -> Self {
        Error::Xml(e)
    }
}


codegen/src/generate.rs
//! XML parsing and code generation functions.

use std::fs::File;
use std::io::{BufWriter, Write};
use std::path::Path;

use amxml::dom::NodePtr;
use chrono::Utc;
use phf_codegen::Map;

use super::Error;

const CATEGORIES_TYPE_SIG: &str = "pub static CATEGORIES: ::phf::Map<&'static str, Category> = ";

/// Contains a builder for a `phf` static hash map.
pub struct CategoryMap(Map<String>);

impl CategoryMap {
    /// Generates a new category hash map from the given root element of the XML DOM.
    ///
    /// Returns `Ok(Self)` if the specification XML was successfully parsed and a static hash map
    /// generated, returns `Err(Error::Xml)` otherwise.
    pub fn generate(root: &NodePtr) -> Result<Self, Error> {
        let mut map = Map::new();

        parse_main_categories(&root, &mut map)?;
        parse_additional_categories(&root, &mut map)?;
        parse_reserved_categories(&root, &mut map)?;

        Ok(CategoryMap(map))
    }

    /// Writes the resulting Rust code to the file located at `out`.
    ///
    /// If the file does not exist at that location, it will automatically be created. Returns
    /// `Ok(())` if successful, returns `Err(Error::Io)` if the file could not be written.
    pub fn write_file(&self, out: &Path) -> Result<(), Error> {
        let CategoryMap(ref map) = *self;

        let mut file = BufWriter::new(File::create(&out)?);
        writeln!(
            file,
            "// Generated by the build script at {}. Do not edit directly!\n",
            Utc::now()
        )?;
        writeln!(file, "/// Static hash map of desktop menu categories.")?;
        write!(file, "{}", CATEGORIES_TYPE_SIG)?;
        map.build(&mut file)?;
        write!(file, ";\n")?;

        Ok(())
    }
}

fn parse_main_categories(root: &NodePtr, map: &mut Map<String>) -> Result<(), Error> {
    let entries = root.get_nodeset(
        "//sect1[@id='main-category-registry']/para/informaltable/tgroup/tbody/row/entry",
    )?;

    let names = entries
        .chunks(3)
        .flat_map(|row| row.into_iter().flat_map(|entry| entry.nth_child(0)).nth(0))
        .map(|name| name.value());

    for name in names {
        let requires = match name.as_str() {
            "Audio" | "Video" => "\"AudioVideo\"",
            _ => "",
        };

        let init_str = format!("Category::Main {{ requires: &[{}] }}", requires);
        map.entry(name, &init_str);
    }

    Ok(())
}

fn parse_additional_categories(root: &NodePtr, map: &mut Map<String>) -> Result<(), Error> {
    let entries = root.get_nodeset(
        "//sect1[@id='additional-category-registry']/para/informaltable/tgroup/tbody/row/entry",
    )?;

    let rows = entries.chunks(3).map(|row| {
        row.into_iter()
            .flat_map(|entry| entry.nth_child(0).map(|text| text.value()))
    });

    for mut row in rows {
        let name = row.next().expect("Couldn't get name column from table row");
        let related_err: String = row.skip(1).collect();
        let related = related_err.replace("QT", "Qt"); // Fix a spelling mistake in the spec.

        let suggests = if related.contains(" or ") {
            related.split(" or ").collect()
        } else if !related.is_empty() {
            vec![related.as_str()]
        } else {
            Vec::new()
        };

        let init_str = format!("Category::Additional {{ suggests: &{:?} }}", suggests);
        map.entry(name, &init_str);
    }

    Ok(())
}

fn parse_reserved_categories(root: &NodePtr, map: &mut Map<String>) -> Result<(), Error> {
    let entries = root.get_nodeset(
        "//sect1[@id='reserved-category-registry']/para/informaltable/tgroup/tbody/row/entry",
    )?;

    let names = entries
        .chunks(2)
        .flat_map(|row| row.into_iter().flat_map(|entry| entry.nth_child(0)).nth(0))
        .map(|name| name.value());

    for name in names {
        map.entry(name, "Category::Reserved");
    }

    Ok(())
}


src/map.rs
// Generated by the build script at 2019-02-15 11:38:15.817512 UTC. Do not edit directly!

/// Static hash map of desktop menu categories.
pub static CATEGORIES: ::phf::Map<&'static str, Category> = ::phf::Map {
    key: 3347381344252206323,
    disps: ::phf::Slice::Static(&[
        (0, 1),
        (0, 11),
        (1, 63),
        (0, 39),
        (17, 66),
        (3, 119),
        (0, 31),
        (1, 21),
        (0, 0),
        (1, 39),
        (1, 0),
        (17, 25),
        (4, 135),
        (0, 2),
        (0, 84),
        (0, 2),
        (0, 0),
        (0, 115),
        (1, 83),
        (3, 26),
        (0, 1),
        (0, 29),
        (5, 140),
        (0, 104),
        (0, 42),
        (44, 29),
        (0, 11),
        (1, 53),
        (31, 106),
    ]),
    entries: ::phf::Slice::Static(&[
        ("ProjectManagement", Category::Additional { suggests: &["Office", "Development"] }),
        ("KDE", Category::Additional { suggests: &["Qt"] }),
        ("Robotics", Category::Additional { suggests: &["Education", "Science"] }),
        ("Profiling", Category::Additional { suggests: &["Development"] }),
        ("Dictionary", Category::Additional { suggests: &["Office", "TextTools"] }),
        ("ArtificialIntelligence", Category::Additional { suggests: &["Education", "Science"] }),
        ("Game", Category::Main { requires: &[] }),
        ("P2P", Category::Additional { suggests: &["Network"] }),
        ("BlocksGame", Category::Additional { suggests: &["Game"] }),
        ("TelephonyTools", Category::Additional { suggests: &["Utility"] }),
        ("TerminalEmulator", Category::Additional { suggests: &["System"] }),
        ("Debugger", Category::Additional { suggests: &["Development"] }),
        ("ComputerScience", Category::Additional { suggests: &["Education", "Science"] }),
        ("Humanities", Category::Additional { suggests: &["Education", "Science"] }),
        ("Shell", Category::Reserved),
        ("GUIDesigner", Category::Additional { suggests: &["Development"] }),
        ("IDE", Category::Additional { suggests: &["Development"] }),
        ("WebDevelopment", Category::Additional { suggests: &["Network", "Development"] }),
        ("RasterGraphics", Category::Additional { suggests: &["Graphics;2DGraphics"] }),
        ("Geoscience", Category::Additional { suggests: &["Education", "Science"] }),
        ("Languages", Category::Additional { suggests: &["Education", "Science"] }),
        ("History", Category::Additional { suggests: &["Education", "Science"] }),
        ("Settings", Category::Main { requires: &[] }),
        ("TV", Category::Additional { suggests: &["AudioVideo;Video"] }),
        ("Music", Category::Additional { suggests: &["AudioVideo", "Education"] }),
        ("Chat", Category::Additional { suggests: &["Network"] }),
        ("Chart", Category::Additional { suggests: &["Office"] }),
        ("CardGame", Category::Additional { suggests: &["Game"] }),
        ("Archiving", Category::Additional { suggests: &["Utility"] }),
        ("Shooter", Category::Additional { suggests: &["Game"] }),
        ("SportsGame", Category::Additional { suggests: &["Game"] }),
        ("Art", Category::Additional { suggests: &["Education", "Science"] }),
        ("Math", Category::Additional { suggests: &["Education", "Science"] }),
        ("Office", Category::Main { requires: &[] }),
        ("Mixer", Category::Additional { suggests: &["AudioVideo;Audio"] }),
        ("System", Category::Main { requires: &[] }),
        ("RemoteAccess", Category::Additional { suggests: &["Network"] }),
        ("Applet", Category::Reserved),
        ("Security", Category::Additional { suggests: &["Settings", "System"] }),
        ("XFCE", Category::Additional { suggests: &["GTK"] }),
        ("Sports", Category::Additional { suggests: &["Education", "Science"] }),
        ("Translation", Category::Additional { suggests: &["Development"] }),
        ("Engineering", Category::Additional { suggests: &[] }),
        ("Documentation", Category::Additional { suggests: &[] }),
        ("Database", Category::Additional { suggests: &["Office", "Development", "AudioVideo"] }),
        ("Spreadsheet", Category::Additional { suggests: &["Office"] }),
        ("Graphics", Category::Main { requires: &[] }),
        ("WordProcessor", Category::Additional { suggests: &["Office"] }),
        ("Printing", Category::Additional { suggests: &["HardwareSettings;Settings"] }),
        ("Java", Category::Additional { suggests: &[] }),
        ("Viewer", Category::Additional { suggests: &["Graphics", "Office"] }),
        ("Astronomy", Category::Additional { suggests: &["Education", "Science"] }),
        ("FileTransfer", Category::Additional { suggests: &["Network"] }),
        ("Tuner", Category::Additional { suggests: &["AudioVideo;Audio"] }),
        ("Maps", Category::Additional { suggests: &["Education", "Science", "Utility"] }),
        ("BoardGame", Category::Additional { suggests: &["Game"] }),
        ("ActionGame", Category::Additional { suggests: &["Game"] }),
        ("Feed", Category::Additional { suggests: &["Network"] }),
        ("2DGraphics", Category::Additional { suggests: &["Graphics"] }),
        ("Utility", Category::Main { requires: &[] }),
        ("Publishing", Category::Additional { suggests: &["Graphics", "Office"] }),
        ("Photography", Category::Additional { suggests: &["Graphics", "Office"] }),
        ("PackageManager", Category::Additional { suggests: &["Settings"] }),
        ("TrayIcon", Category::Reserved),
        ("Education", Category::Main { requires: &[] }),
        ("News", Category::Additional { suggests: &["Network"] }),
        ("LogicGame", Category::Additional { suggests: &["Game"] }),
        ("Spirituality", Category::Additional { suggests: &["Education", "Science", "Utility"] }),
        ("HardwareSettings", Category::Additional { suggests: &["Settings"] }),
        ("FileTools", Category::Additional { suggests: &["Utility", "System"] }),
        ("Sequencer", Category::Additional { suggests: &["AudioVideo;Audio"] }),
        ("Screensaver", Category::Reserved),
        ("Compression", Category::Additional { suggests: &["Utility;Archiving"] }),
        ("GTK", Category::Additional { suggests: &[] }),
        ("Building", Category::Additional { suggests: &["Development"] }),
        ("DiscBurning", Category::Additional { suggests: &["AudioVideo"] }),
        ("StrategyGame", Category::Additional { suggests: &["Game"] }),
        ("TextEditor", Category::Additional { suggests: &["Utility"] }),
        ("Development", Category::Main { requires: &[] }),
        ("Economy", Category::Additional { suggests: &["Education", "Science"] }),
        ("Adult", Category::Additional { suggests: &[] }),
        ("MedicalSoftware", Category::Additional { suggests: &["Education", "Science"] }),
        ("Scanning", Category::Additional { suggests: &["Graphics"] }),
        ("Geography", Category::Additional { suggests: &["Education", "Science"] }),
        ("Physics", Category::Additional { suggests: &["Education", "Science"] }),
        ("TextTools", Category::Additional { suggests: &["Utility"] }),
        ("Telephony", Category::Additional { suggests: &["Network"] }),
        ("ArcadeGame", Category::Additional { suggests: &["Game"] }),
        ("Finance", Category::Additional { suggests: &["Office"] }),
        ("Audio", Category::Main { requires: &["AudioVideo"] }),
        ("KidsGame", Category::Additional { suggests: &["Game"] }),
        ("HamRadio", Category::Additional { suggests: &["Network", "Audio"] }),
        ("Chemistry", Category::Additional { suggests: &["Education", "Science"] }),
        ("Presentation", Category::Additional { suggests: &["Office"] }),
        ("DesktopSettings", Category::Additional { suggests: &["Settings"] }),
        ("FileManager", Category::Additional { suggests: &["System;FileTools"] }),
        ("ParallelComputing", Category::Additional { suggests: &["Education;ComputerScience", "Science;ComputerScience"] }),
        ("Recorder", Category::Additional { suggests: &["Audio", "Video", "AudioVideo"] }),
        ("Literature", Category::Additional { suggests: &["Education", "Science"] }),
        ("Amusement", Category::Additional { suggests: &[] }),
        ("Dialup", Category::Additional { suggests: &["Network"] }),
        ("Biology", Category::Additional { suggests: &["Education", "Science"] }),
        ("ImageProcessing", Category::Additional { suggests: &["Education", "Science"] }),
        ("Electronics", Category::Additional { suggests: &[] }),
        ("Player", Category::Additional { suggests: &["Audio", "Video", "AudioVideo"] }),
        ("IRCClient", Category::Additional { suggests: &["Network"] }),
        ("Video", Category::Main { requires: &["AudioVideo"] }),
        ("NumericalAnalysis", Category::Additional { suggests: &["Education;Math", "Science;Math"] }),
        ("Midi", Category::Additional { suggests: &["AudioVideo;Audio"] }),
        ("PDA", Category::Additional { suggests: &["Office"] }),
        ("Construction", Category::Additional { suggests: &[] }),
        ("FlowChart", Category::Additional { suggests: &["Office"] }),
        ("Qt", Category::Additional { suggests: &[] }),
        ("ConsoleOnly", Category::Additional { suggests: &[] }),
        ("Simulation", Category::Additional { suggests: &["Game"] }),
        ("ContactManagement", Category::Additional { suggests: &["Office"] }),
        ("Core", Category::Additional { suggests: &[] }),
        ("Email", Category::Additional { suggests: &["Office", "Network"] }),
        ("Network", Category::Main { requires: &[] }),
        ("Electricity", Category::Additional { suggests: &["Education", "Science"] }),
        ("Accessibility", Category::Additional { suggests: &["Settings", "Utility"] }),
        ("Calculator", Category::Additional { suggests: &["Utility"] }),
        ("AudioVideoEditing", Category::Additional { suggests: &["Audio", "Video", "AudioVideo"] }),
        ("VideoConference", Category::Additional { suggests: &["Network"] }),
        ("Filesystem", Category::Additional { suggests: &["System"] }),
        ("Emulator", Category::Additional { suggests: &["System", "Game"] }),
        ("Calendar", Category::Additional { suggests: &["Office"] }),
        ("Clock", Category::Additional { suggests: &["Utility"] }),
        ("RolePlaying", Category::Additional { suggests: &["Game"] }),
        ("Geology", Category::Additional { suggests: &["Education", "Science"] }),
        ("RevisionControl", Category::Additional { suggests: &["Development"] }),
        ("Motif", Category::Additional { suggests: &[] }),
        ("Science", Category::Main { requires: &[] }),
        ("GNOME", Category::Additional { suggests: &["GTK"] }),
        ("AudioVideo", Category::Main { requires: &[] }),
        ("3DGraphics", Category::Additional { suggests: &["Graphics"] }),
        ("DataVisualization", Category::Additional { suggests: &["Education", "Science"] }),
        ("WebBrowser", Category::Additional { suggests: &["Network"] }),
        ("VectorGraphics", Category::Additional { suggests: &["Graphics;2DGraphics"] }),
        ("Monitor", Category::Additional { suggests: &["System", "Network"] }),
        ("OCR", Category::Additional { suggests: &["Graphics;Scanning"] }),
        ("AdventureGame", Category::Additional { suggests: &["Game"] }),
        ("InstantMessaging", Category::Additional { suggests: &["Network"] }),
    ]),
};
